_posts/2022-09-15-redis-job-queue.adoc

```diff
@@ -0,0 +1,302 @@
+The simulators send the outcome to a _channel_ consumed by the application.
+The application then broadcasts these outcomes to a web page.
+
+This post only discusses the interaction with Redis.
+The rest of the application is straightforward and just uses RESTEasy Reactive and Hibernate ORM with Panache.
+You can find the full code of the application on https://github.com/cescoffier/quarkus-redis-job-queue-demo.
+
+== Submitting jobs
+
+The first task is to modelize the job queue.
+We are using a https://redis.io/docs/data-types/#lists[Redis list] to store the https://github.com/cescoffier/quarkus-redis-job-queue-demo/blob/main/fight-simulator/src/main/java/me/escoffier/quarkus/redis/fight/FightRequest.java[FightRequest].
+
+[source,java]
+----
+package me.escoffier.quarkus.redis.fight;
+
+public record FightRequest(String id, Hero hero, Villain villain) {
+
+}
+----
+
+Redis lists distinguish the left side of the list from the right side of the list.
+This distinction allows implementing a FIFO queue where we write on the left side and consume from the right side.
+
+To manipulate a Redis list, we need the _group of commands_ associated with this data structure.
+In the https://github.com/cescoffier/quarkus-redis-job-queue-demo/blob/main/supes-application/src/main/java/me/escoffier/quarkus/redis/supes/SupesService.java[SupesService class], we inject the `RedisDataSource` and retrieve the group of commands:
+
+[source,java]
+----
+public SupesService(RedisDataSource dataSource, ...) {
+    commands = dataSource.list(FightRequest.class);
+  // ...
+}
+----
+
+Let's now look at the `submitAFight` method:
+
+[source,java]
+----
+public FightRequest submitAFight() {
+    var hero = Hero.getRandomHero();
+    var villain = Villain.getRandomVillain();
+    var id = UUID.randomUUID().toString();
+    var request = new FightRequest(id, hero, villain);
+    commands.lpush("fight-requests", request);
+    return request;
+}
+----
+
+The `submitaFight` method retrieves the random fighters, computes an id,  builds the `FightRequest` instance, and executes the `LPUSH` command.
+The `LPUSH`  command writes the given item to the left side of the list stored at the given key (`fight-requests`).
+
+== Receiving the job requests
+
+Let's now look at the other side: the fight simulators.
+The simulators poll the `FightRequests` from the Redis list modelizing our job queue and simulate the fight.
+
+The simulator is implemented in https://github.com/cescoffier/quarkus-redis-job-queue-demo/blob/main/fight-simulator/src/main/java/me/escoffier/quarkus/redis/fight/FightSimulator.java[`me.escoffier.quarkus.redis.fight.FightSimulator`].
+The constructor receives a configured name (to distinguish multiple simulators) and the Redis data source.
+It creates the objects to emit the Redis commands to read from a Redis list:
+
+[source,java]
+----
+public FightSimulator(@ConfigProperty(name = "simulator-name") String name, Logger logger, RedisDataSource ds) {
```

Comment:
Logger in the constructor? 
It's valid, but it's not really something use often :)

Reply:
Initially, I was logging in to the constructor, which requires constructor injection, because the field was not yet injected.